library/psa_crypto.c

```diff
@@ -3708,39 +3709,34 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
     return( status );
 }
 
-static psa_status_t psa_validate_tag_length( psa_aead_operation_t *operation,
-                                             psa_algorithm_t alg ) {
-    uint8_t tag_len = 0;
-    if( psa_driver_get_tag_len( operation, &tag_len ) != PSA_SUCCESS )
-    {
-        return( PSA_ERROR_INVALID_ARGUMENT );
-    }
+static psa_status_t psa_validate_tag_length( psa_algorithm_t alg ) {
+    const uint8_t tag_len = PSA_ALG_AEAD_GET_TAG_LENGTH( alg );
 
     switch( PSA_ALG_AEAD_WITH_SHORTENED_TAG( alg, 0 ) )
     {
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)
+#if defined(PSA_WANT_ALG_CCM)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG( PSA_ALG_CCM, 0 ):
             /* CCM allows the following tag lengths: 4, 6, 8, 10, 12, 14, 16.*/
             if( tag_len < 4 || tag_len > 16 || tag_len % 2 )
                 return( PSA_ERROR_INVALID_ARGUMENT );
             break;
-#endif /* MBEDTLS_PSA_BUILTIN_ALG_CCM */
+#endif /* PSA_WANT_ALG_CCM */
 
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)
+#if defined(PSA_WANT_ALG_GCM)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG( PSA_ALG_GCM, 0 ):
             /* GCM allows the following tag lengths: 4, 8, 12, 13, 14, 15, 16. */
             if( tag_len != 4 && tag_len != 8 && ( tag_len < 12 || tag_len > 16 ) )
                 return( PSA_ERROR_INVALID_ARGUMENT );
             break;
-#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#endif /* PSA_WANT_ALG_GCM */
 
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+#if defined(PSA_WANT_ALG_CHACHA20_POLY1305)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG( PSA_ALG_CHACHA20_POLY1305, 0 ):
             /* We only support the default tag length. */
             if( tag_len != 16 )
```

Comment:
If this is a limitation of our implementation and `PSA_WANT_ALG_CHACHA20_POLY1305` do support other tag lengths (I did not check if this is the case but the comment goes in that direction) this check should not be there but in psa_crypto_aead.c. Here we should check the tag lengths allowed by the specification of CHACHA20_POLY1305. At least we have to change the comment here.

Reply:
From the specification:
> Poly1305 is a one-time authenticator designed by D. J. Bernstein.
   Poly1305 takes a 32-byte one-time key and a message and produces a
   16-byte tag.  This tag is used to authenticate the message.

https://datatracker.ietf.org/doc/html/rfc7539

Reply:
ok thus I guess we just have to change the comment "We only support the default tag length." that is misleading.

Reply:
The tag length is encoded in the algorithm value. 16 is the tag length from `PSA_ALG_CHACHA20_POLY1305`. That's the default tag length, as opposed to other tag lengths which you can specify with the help of `PSA_ALG_AEAD_WITH_SHORTENED_TAG`.

RFC 7539 only defines AEAD_CHACHA20_POLY1305 with a 16-byte tag, but that does not prevent others from defining a truncated version where the tag is simply truncated. The PSA API recommends not truncating the tag, but does not forbid it (in case there's a network protocol that does this — IIRC that's not the case for TLS, not for chacha-poly, but it could be the case for other protocols).

Currently, Mbed TLS only allows a 16-byte tag and should keep filtering the tag length in the core, independently of whether drivers might allow shorter tags.

Reply:
I understand that the comment make sense, based on @gilles-peskine-arm comment.